### ISSUE TITLE
no_default_cases seems strange

### ISSUE THREAD

**Issue by Hixie** [dart-lang/linter#4735]
I don't understand the value of `no_default_cases` and its message is confusing.

First the message. Right now it just says "Invalid use of 'default' member in a switch". However, this is highly misleading, because as far as I can tell the syntax is perfectly fine, it's just that the lint wants to discourage the use of `default` with enums. So I think a message more like "Use of 'default' in a switch for an enum is discouraged" or something (maybe with a different message for enums and enum-like classes).

More importantly, I don't understand the value of this lint. I ran into it because I have a case where I have an enum with four values, which might have new values added in the future, and precisely two of them need special handling. I could do an if-else-if chain to compare the two values, but that seems like exactly what `switch` is for. But I don't want to have to specify all the enum values for the no-op case, I just want to say `default: // do nothing` or similar. Why is this not a valid use?

**Comment by pq** [comment]
Some rationale here: https://github.com/dart-lang/linter/issues/2084

**Comment by Hixie** [comment]
The only rationale I see in that issue is that it "causes bugs", did I miss something else?

Why does it "cause bugs"? What bugs is it causing?

I certainly agree that if you have covered every case you don't want to also give a `default`, but I'm pretty sure dead code checking will catch that. I could see an argument that the enum should complain about `default` if there is only one value that's been missed, though that's debatable. But I don't understand what the rationale is for the general case.

**Comment by pq** [comment]
Nah. It was motivated by an (internal) team that just wanted to disallow all default handling. I can't recall any more specifics.

I wouldn't enable it myself...

**Comment by Hixie** [comment]
I think it would be useful if the lint description said something to that effect (or more to the point, something like "this isn't necessarily best practice"). The problem with the current documentation is that it very confidently asserts what's GOOD and what's BAD and someone without context might think that's because the dart style guide says it or something.

(This is something I've seen in other lint descriptions too. Maybe we need a general audit of the lint documentations to make sure they all provide appropriate context throughout.)

**Comment by pq** [comment]
I agree 💯 on all counts. We've talked about a big documentation fixit but haven't been able to prioritize it.

/fyi @parlough @bwilkerson @MaryaBelanger

**Comment by stuartmorgan** [comment]
> More importantly, I don't understand the value of this lint.

The argument I've seen for checks like this in other languages is to guarantee getting compiler help for future maintenance (at the expense of added verbosity).

> I ran into it because I have a case where I have an enum with four values, which might have new values added in the future, and precisely two of them need special handling.

As you said in your case, new values might be added in the future. *Right now* precisely two of the four need special handling, but will the fifth value need special handling here? If you write it with a `default:`, you've permanently disabled compiler help for that scenario, and who knows if someone will remember to come look at your case (or be able to find it) when adding a new value, whereas if you explicitly list the two that don't need any handling then when someone adds a fifth value the compiler will force them to consider what should happen in that switch and make a decision one way or the other.

(Whether you think that tradeoff is a good one is, like most lints, a matter of opinion, but that's the value it provides.)

**Comment by Hixie** [comment]
@stuartmorgan I get that in principle, and that's why https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values, but banning it entirely seems like an odd choice.

In any case I guess the real complaint here is just about the message and documentation, not the lint's existence.

**Comment by stuartmorgan** [comment]
> but banning it entirely seems like an odd choice

Lints don't ban things entirely; there's `// ignore`.

(Honestly given that style entry, this seems like the lint is working exactly as intended: the style guide for Flutter says to *not* do exactly what you are doing here, and to instead do what the lint suggests doing, and the linter is helping enforce that.)

**Comment by Hixie** [comment]
By "banning it entirely" I mean the wording calling it "invalid" and the documentation just saying it's "BAD" without justification. (As opposed to the style guide which gives a reason and uses the word "avoid".)

**Comment by jakemac53** [comment]
Yeah, it seems like a message such as `Avoid "default" cases in switches on enums` would be a lot better - that indicates a lot more clearly that this is a suggestion, not a requirement.
